File: python/examples/02_execute.py
"""
Example 02

Execute a Business Transaction.
"""

from __future__ import annotations

import json
from dataclasses import asdict
from datetime import UTC, datetime
from uuid import uuid4

from parmana import (
    Authority,
    Authorization,
    BusinessTransaction,
    BusinessTransactionMetadata,
    Intent,
    ParmanaClient,
    PolicyReference,
)


def main() -> None:
    client = ParmanaClient(
        endpoint="http://localhost:3000",
    )

    #
    # Generate a unique transaction ID for every execution.
    #
    transaction_id = str(uuid4())

    #
    # Use timezone-aware UTC timestamps.
    #
    now = datetime.now(UTC)

    transaction = BusinessTransaction(
        business_transaction_id=transaction_id,

        metadata=BusinessTransactionMetadata(
            business_transaction_id=transaction_id,
            correlation_id="demo-execution",
            tenant_id=None,
            source_system="python-sdk-example",
            submitted_by="sdk-demo",
            submitted_at=now,
        ),

        authority=Authority(
            authority_id="authority-001",
            authority_type="SERVICE",
            principal_id="python-sdk",
            display_name="Python SDK",
            issued_at=now,
        ),

        authorization=Authorization(
            authorization_id="authorization-001",
            authority_id="authority-001",
            purpose="Execute demo transaction",
            issued_at=now,
        ),

        intent=Intent(
            intent_id="intent-001",
            authorization_id="authorization-001",
            action="VendorPayment",
            target="vendor/V-100",
            parameters={
                "amount": 1000,
                "currency": "USD",
            },
            created_at=now,
        ),

        policy=PolicyReference(
            name="vendor-payment",
            version="1.0.0",
            schema_version="1.0.0",
        ),

        signals={
            "vendorVerified": True,
            "paymentApproved": True,
            "amount": 1000,
        },

        status="RECEIVED",

        created_at=now,
    )

    trust_record = client.execution.execute(transaction)

    print(
        json.dumps(
            asdict(trust_record),
            indent=2,
            default=str,
        )
    )


if __name__ == "__main__":
    main()

File: typescript/examples/02-execute.ts
import { pathToFileURL } from "node:url";

import {
  BusinessTransaction,
  HttpTransport,
  ParmanaClient,
} from "@parmana/sdk";

/**
 * Constructs a ParmanaClient, submits a Business Transaction through
 * POST /execute, and returns the resulting Execution Trust Record.
 * Exported (rather than only run as a top-level script) so
 * typescript/test/integration/examples.integration.test.ts can prove
 * this exact example actually works against a real running server,
 * not just that it compiles.
 */
export async function runExecuteExample(
  endpoint = "http://localhost:3000",
) {
  const client =
    new ParmanaClient({
      endpoint,

      transport:
        new HttpTransport({
          endpoint,
        }),
    });

  const businessTransactionId =
    crypto.randomUUID();

  const transaction: BusinessTransaction = {
    businessTransactionId,

    metadata: {
      businessTransactionId,

      correlationId:
        "corr-001",

      tenantId:
        "tenant-001",

      sourceSystem:
        "typescript-sdk-example",

      submittedBy:
        "demo-user",

      submittedAt:
        new Date(),
    },

    authority: {
      authorityId:
        "authority-001",

      authorityType:
        "USER",

      principalId:
        "alice@example.com",

      displayName:
        "Alice",

      issuedAt:
        new Date(),
    },

    authorization: {
      authorizationId:
        "authorization-001",

      authorityId:
        "authority-001",

      purpose:
        "Vendor payment approval",

      issuedAt:
        new Date(),
    },

    intent: {
      intentId:
        "intent-001",

      authorizationId:
        "authorization-001",

      // test:fixture-execute, not TransferFunds/payments:execute: the
      // capability this example originally targeted was never
      // registered by any connector in this repository (a NODE_ENV=test
      // -only generic fixture connector plays that zero-external
      // -dependency role instead, see
      // createTestFixtureConnector.ts and docs/VERIFICATION-GAPS.md
      // G-27) -- so this example, never covered by any test or build
      // (excluded from typescript/tsconfig.json's own include list),
      // had never actually been run against a real server before this
      // pass proved it.
      action:
        "test:fixture-execute",

      target:
        "vendor/vendor-123",

      parameters: {
        amount: 100,
        currency: "USD",
      },

      createdAt:
        new Date(),
    },

    policy: {
      name:
        "vendor-payment",

      version:
        "2.0.0",

      schemaVersion:
        "1.0.0",
    },

    signals: {
      vendorVerified: true,
      invoiceVerified: true,
      paymentApproved: true,
      sufficientFunds: true,
      paymentAmount: 100,
      riskScore: 5,
      // vendor-payment@2.0.0 declares boundSignals: vendorId -> target;
      // SignalIntentBinder rejects this transaction unless this signal
      // exactly equals intent.target, checked before policy evaluation
      // ever runs (docs/VERIFICATION-GAPS.md G-24).
      vendorId: "vendor/vendor-123",
    },

    status:
      "RECEIVED",

    createdAt:
      new Date(),
  };

  return client.execute(transaction);
}

const isMainModule =
  process.argv[1] !== undefined &&
  import.meta.url === pathToFileURL(process.argv[1]).href;

if (isMainModule) {
  const trustRecord =
    await runExecuteExample(
      process.env.PARMANA_EXAMPLE_ENDPOINT,
    );

  console.log(
    JSON.stringify(
      trustRecord,
      null,
      2,
    ),
  );
}
